sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc

```diff
@@ -0,0 +1,457 @@
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+SYCL 2020 provides multiple ways to enqueue work to a device. In some cases, a
+single function name is used to enqueue kernels with very different use-cases
+and execution models (e.g., `parallel_for(range)` and
+`parallel_for(nd_range)`). In almost all cases, the functions are available in
+multiple places (e.g., `queue::parallel_for` and `handler::parallel_for`).
+In all cases, these functions return an `event` object by default, which has
+been shown to introduce undesirable performance overhead.
+
+This extension addresses these issues by:
+
+1. Using different function names for different use-cases.
+2. Using free-functions instead of member functions.
+3. Requiring developers to opt-in to the creation of `event` objects.
+
+This extension makes SYCL simpler and easier to document. It is also expected
+to improve the performance of many SYCL applications, where `event` objects are
+not required to describe application behavior.
+
+All functions proposed in this extension accept as their first argument an
+object that represents where a command should be submitted. For now, we refer
+to this as an "Executor". This argument can currently be either a
+`sycl::handler` or a `sycl::queue`, allowing the new functions to be used
+either at command-group scope or as a replacement for existing queue shortcuts.
+It is expected that a future version of this extension will adjust this
+overload set to include functions compatible with future C++ concepts (such as
+senders and receivers).
+
+
+=== Usage example
+
+The example below demonstrates that the syntax proposed here requires only
+minor changes to existing applications, while retaining their structure.
+
+
+==== SYCL 2020
+
+[source,c++]
+----
+q.submit([&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  h.parallel_for(1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e = q.parallel_for(sycl::nd_range<1>{1024, 16}, depEvents,
+  [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+});
+e.wait();
+----
+
+
+==== Proposed syntax
+
+[source,c++]
+----
+using syclex = sycl::ext::oneapi::experimental;
+
+syclex::submit(q, [&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  syclex::parallel_for(h, 1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e;
+syclex::nd_launch(q, sycl::nd_range<1>{1024, 16}, [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+}, depEvents, e);
```

Comment:
Two possible fixes here:

1. I think this needs to pass the address of the SYCL event, not the event itself?
2. To have the same semantics as the SYCL 2020 example above we need to wait on the event.

```suggestion
}, depEvents, &e);
e.wait();
```